### PR TITLE
Commiter Work - Object Explorer

### PR DESCRIPTION
**Pull request by abist**
Changes made -
- [x] Cleaned up copyright header for all files
- [x] Added new extension viewlet for queries and Object Explorer
- [x] Added existing `New Query` command as context menu action for object explorer
- [x] Added Object Explorer as MRU

Screenshots - 

EDIT: Updated Screenshot with Icons
Untitled

### CODE REVIEW

File: src/constants/constants.ts [hunk 1]
```diff
@@ -1,3 +1,8 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
```
Comment (Charles-Gagnon) [inline src/constants/constants.ts]: It'd be great to add a gulp hygiene task like we have for ADS that checks for these as a precommit hook

Comment (kevcunnane) [reply src/constants/constants.ts]: We do - checking in changes to a file will flag missing copyright (and just a space before it, as I've discovered a bunch).

Comment (kevcunnane) [reply src/constants/constants.ts]: Oh sorry - wrong repo. Ignore my comment.

File: src/controllers/localWebService.ts [hunk 2]
```diff
@@ -43,8 +48,10 @@ export default class LocalWebService {
         this.server.on('request', this.app);
 
         // Handle new connections to the web socket server
-        this.wss.on('connection', (ws) => {
-            let parse: any = querystring.parse(<string>url.parse(ws.upgradeReq.url).query);
+        this.wss.on('connection', (ws, req) => {
+            // upgradeReq was removed from the API
+            // See - https://github.com/websockets/ws/pull/1099
+            let parse: any = querystring.parse(<string>url.parse(req.url).query);
```
Comment (Charles-Gagnon) [inline src/controllers/localWebService.ts]: Was this actually broken? Or did you just happen to see that it was removed and did this now so that if we updated the version later on it wouldn't break? Because right now it seems like this isn't actually changing/fixing anything

Comment (abist) [reply src/controllers/localWebService.ts]: No it was broken in master when I started working on this. And I saw the last commit was bumping the version for this library.

File: package.nls.json [hunk 3]
```diff
@@ -2,6 +2,10 @@
 "extension.runQuery":"Execute Query",
 "extension.runCurrentStatement":"Execute Current Statement",
 "extension.cancelQuery":"Cancel Query",
+"extension.addObjectExplorer":"Add Object Explorer",
+"extension.removeObjectExplorerNode":"Remove",
```
Comment (Charles-Gagnon) [inline package.nls.json]: Wouldn't something like "Show Object Explorer" be a better title? Or am I misunderstanding what this command is doing?

Comment (abist) [reply package.nls.json]: Well, there's a `View: Object Explorer` already that we get for free since I added a new `Viewlet`/`View` for Object Explorer. This command basically adds a node to the tree for a new server.

Comment (Charles-Gagnon) [reply package.nls.json]: Then shouldn't it be Add Server or Add Connection or something? Right now it's implying that it's adding "Object Explorer" (whatever that means) so it seems like this could be better clarified

---
In reply to: [303570661](https://github.com/microsoft/vscode-mssql/pull/1259#discussion_r303570661) [](ancestors = 303570661)

File: package.json [hunk 4]
```diff
@@ -123,6 +126,23 @@
     "ws": ">=3.3.1"
   },
   "contributes": {
+    "viewsContainers": {
+      "activitybar": [
+        {
+          "id": "objectExplorer",
+          "title": "Object Explorer",
+          "icon": "./images/server_page_inverse.svg"
+        }
```
Comment (Charles-Gagnon) [inline package.json]: Does this not support light/dark images?

Comment (Charles-Gagnon) [reply package.json]: [https://code.visualstudio.com/api/references/contribution-points#contributes.viewsContainers](https://code.visualstudio.com/api/references/contribution-points#contributes.viewsContainers) Looks like not - just double check this to be sure though if you wouldn't mind

---
In reply to: [303623862](https://github.com/microsoft/vscode-mssql/pull/1259#discussion_r303623862) [](ancestors = 303623862)

File: package.nls.json (hunk 3 above)
Comment (abist) [reply package.nls.json]: @Charles-Gagnon Sure, I might need to consult Alan once he's back for certain strings.

File: src/controllers/localWebService.ts (hunk 2 above)
Comment (Charles-Gagnon) [reply src/controllers/localWebService.ts]: Alright. I'd just remove the comment though. It has no context here unless you know the history (that it used to be ws.upgradeReq). The logic just using req.url seems fine by itself - not sure why it was using upgradeReq in the first place.

---
In reply to: [303567328](https://github.com/microsoft/vscode-mssql/pull/1259#discussion_r303567328) [](ancestors = 303567328)

File: src/objectExplorer/objectExplorerProvider.ts [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+import * as vscode from 'vscode';
+import ConnectionManager from '../controllers/connectionManager';
+import { NodeInfo } from '../models/contracts/objectExplorer/nodeInfo';
+import { ObjectExplorerService } from './objectExplorerService';
+import { ConnectionCredentials } from '../models/connectionCredentials';
+
+export class TreeNodeInfo extends vscode.TreeItem {
```
Comment (Charles-Gagnon) [inline src/objectExplorer/objectExplorerProvider.ts]: >TreeNodeInfo [](start = 13, length = 12)

I'd suggest putting this into its own file - no reason to overload this file with additional stuff

File: src/objectExplorer/objectExplorerService.ts [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+                return this._treeNodeToChildrenMap.get(this._currentNode);
+            } else {
+                // expansion
+                await this.expandNode(element, element.sessionId);
+                return;
+            }
+        } else {
+            if (this._rootTreeNodeArray.length === 0) {
+                this.createSession();
+                return;
```
Comment (Charles-Gagnon) [inline src/objectExplorer/objectExplorerService.ts]: >return [](start = 16, length = 6)

You should return an empty array here and above

File: src/controllers/mainController.ts [hunk 7]
```diff
@@ -113,10 +115,38 @@ export default class MainController implements vscode.Disposable {
...
+                    vscode.window.registerTreeDataProvider('objectExplorer', this._objectExplorerProvider)
+                );
+                this.registerCommand(Constants.cmdAddObjectExplorer);
+                this._event.on(Constants.cmdAddObjectExplorer, async () => {
+                    if (self._objectExplorerProvider.objectExplorerExists) {
+                        return self._objectExplorerProvider.createSession();
+                    } else {
+                        self._objectExplorerProvider.objectExplorerExists = true;
+                        return vscode.commands.executeCommand('workbench.view.extension.objectExplorer');
+                    }
```
Comment (Charles-Gagnon) [inline src/controllers/mainController.ts]: This command ID should go with the rest of the ones in constants.ts I would think. Also I'm not seeing this registered anywhere?

File: src/controllers/mainController.ts [hunk 8]
```diff
@@ -113,10 +115,38 @@ export default class MainController implements vscode.Disposable {
...
+                this._context.subscriptions.push(
+                    vscode.window.registerTreeDataProvider('objectExplorer', this._objectExplorerProvider)
+                );
+                this.registerCommand(Constants.cmdAddObjectExplorer);
+                this._event.on(Constants.cmdAddObjectExplorer, async () => {
+                    if (self._objectExplorerProvider.objectExplorerExists) {
+                        return self._objectExplorerProvider.createSession();
+                    } else {
+                        self._objectExplorerProvider.objectExplorerExists = true;
+                        return vscode.commands.executeCommand('workbench.view.extension.objectExplorer');
```
Comment (Charles-Gagnon) [inline src/controllers/mainController.ts]: This seems odd to have it set here. Shouldn't the objectExplorerProvider be the one setting this whenever it's ready?

File: src/controllers/mainController.ts (hunk 7 above)
Comment (abist) [reply src/controllers/mainController.ts]: Every `view` that is registered with VSCode automatically gets a `workbench.view.extension.view_name` command that opens it up.

File: src/controllers/mainController.ts (hunk 8 above)
Comment (Charles-Gagnon) [reply src/controllers/mainController.ts]: What does this even control anyways? I'm not sure exists is the right term here - it seems like this is more saying whether the provider is ready or not?

---
In reply to: [303634495](https://github.com/microsoft/vscode-mssql/pull/1259#discussion_r303634495) [](ancestors = 303634495)

Comment (abist) [reply src/controllers/mainController.ts]: This is just for the case when a user opens the `Object Explorer` for the first time. The provider doesn't know that yet.

File: src/controllers/mainController.ts [hunk 9]
```diff
@@ -499,11 +529,22 @@ export default class MainController implements vscode.Disposable {
...
+                this._untitledSqlDocumentService.newQuery().then((uri) => {
+                    // connect to the node if the command came from the context
+                    if (!this.connectionManager.isConnected(sessionId)) {
+                        const connectionCreds = this._objectExplorerProvider.getConnectionCredentials(sessionId);
+                        this._statusview.languageFlavorChanged(uri.toString(), Constants.mssqlProviderName);
+                        return this.connectionManager.connect(uri.toString(), connectionCreds);
+                    }
+                });
+            } else {
+                return this._untitledSqlDocumentService.newQuery().then(x => {
```
Comment (Charles-Gagnon) [inline src/controllers/mainController.ts]: >x [](start = 72, length = 1)

suggest updating this to uri as well to be clear what it actually is

**Comment by Charles-Gagnon** [comment]
So is this only going to display active connections then? That is if I make some connections and then close VS Code the next time I open it up those connections won't show up in OE until you connect to them again?

File: package.json [hunk 10]
```diff
@@ -154,6 +174,35 @@
...
+        {
+          "command": "extension.objectExplorerNewQuery",
+          "when": "viewItem == Server"
+        },
+        {
+          "command": "extension.objectExplorerNewQuery",
+          "when": "viewItem == Database"
+        },
+        {
+          "command": "extension.objectExplorerNewQuery",
```
Comment (kburtram) [inline package.json]: if we could simply add "SELECT TOP 1000" here that would be awesome.  That's one of the top use-cases users want to complete when viewing a table in OE.

File: package.nls.json (hunk 3 above)
Comment (kburtram) [reply package.nls.json]: yeah, let's use the same text we use in ADS for this purpose: "Add Connection".

File: test/mainController.test.ts [hunk 11]
```diff
@@ -199,21 +204,21 @@ suite('MainController Tests', () => {
 
     test('onNewQuery should call the new query and new connection' , () => {
 
-        untitledSqlDocumentService.setup(x => x.newQuery()).returns(() => Promise.resolve(true));
+        untitledSqlDocumentService.setup(x => x.newQuery()).returns(() => Promise.resolve(undefined));
```
Comment (kburtram) [inline test/mainController.test.ts]: can we add some tests for these changes?

File: src/objectExplorer/objectExplorerService.ts [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+                return this._treeNodeToChildrenMap.get(this._currentNode);
+            } else {
+                // expansion
+                await this.expandNode(element, element.sessionId);
+                return [];
+            }
+        } else {
+            if (this._rootTreeNodeArray.length === 0) {
+                this.createSession();
+                return;
```
Comment (Charles-Gagnon) [inline src/objectExplorer/objectExplorerService.ts]: >return [](start = 16, length = 6)

Still should be returning an empty array here

File: src/controllers/mainController.ts [hunk 13]
```diff
@@ -113,10 +113,51 @@ export default class MainController implements vscode.Disposable {
...
+                this._event.on(Constants.cmdRemoveObjectExplorerNode, () => {
+                    return this._objectExplorerProvider.removeObjectExplorerNode(this._objectExplorerProvider.currentNode);
+                });
+                this.registerCommand(Constants.cmdRefreshObjectExplorerNode);
+                this._event.on(Constants.cmdRefreshObjectExplorerNode, () => {
+                    return this._objectExplorerProvider.refreshNode(this._objectExplorerProvider.currentNode);
+                });
+                this._context.subscriptions.push(vscode.commands.registerCommand(Constants.cmdScriptSelect, async (treeNodeInfo) => {
+                    const objectNames = treeNodeInfo.label.split('.');
+                    const tableName = `[${objectNames[0]}].[${objectNames[1]}]`;
```
Comment (Charles-Gagnon) [inline src/controllers/mainController.ts]: >onst tableName = `[${objectNames[0]}].[${objectNames[1]}]` [](start = 21, length = 58)

These names need to be escaped

File: src/controllers/vscodeWrapper.ts [hunk 14]
```diff
@@ -169,8 +174,8 @@ export default class VscodeWrapper {
      * @return A promise that resolves to a [document](#TextDocument).
      * @see vscode.workspace.openTextDocument
      */
-    public openMsSqlTextDocument(): Thenable<vscode.TextDocument> {
-        return vscode.workspace.openTextDocument({ language: 'sql'});
+    public openMsSqlTextDocument(content?: string): Thenable<vscode.TextDocument> {
```
Comment (Charles-Gagnon) [inline src/controllers/vscodeWrapper.ts]: >content [](start = 33, length = 7)

Update doc comments

File: test/mainController.test.ts (hunk 11 above)
Comment (Charles-Gagnon) [reply test/mainController.test.ts]: Yes, please add some tests

---
In reply to: [305093888](https://github.com/microsoft/vscode-mssql/pull/1259#discussion_r305093888) [](ancestors = 305093888)

File: src/models/contracts/objectExplorer/refreshSessionRequest.ts [hunk 15]
```diff
@@ -0,0 +1,22 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+import { ExpandParams } from './expandNodeRequest';
+import { RequestType } from 'vscode-jsonrpc';
+
+/**
+ * Parameters to the ExpandRequest.
```
Comment (Charles-Gagnon) [inline src/models/contracts/objectExplorer/refreshSessionRequest.ts]: >ExpandRequest [](start = 21, length = 13)

RefreshRequest

File: src/models/contracts/objectExplorer/refreshSessionRequest.ts [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+ */
+export class RefreshParams extends ExpandParams {}
```
Comment (Charles-Gagnon) [inline src/models/contracts/objectExplorer/refreshSessionRequest.ts]: >ExpandParams [](start = 35, length = 12)

I don't think this should be extending ExpandParams. They have nothing to do with each other so shouldn't be directly related. It's fine in this case to just copy the params if you want to use the same ones

File: src/models/contracts/objectExplorer/nodeInfo.ts [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+     * The path functions as a unique ID and is used to disambiguate the node when sending requests for expansion.
+     * A common ID is needed since processes do not share address space and need a unique identifier
+     */
+    public nodePath: string;
+
+    /**
+     * The type of node - for example Server, Database, Folder, Table
+     */
+    public nodeType: string;
+
```
Comment (Charles-Gagnon) [inline src/models/contracts/objectExplorer/nodeInfo.ts]: How is this set? It would be greatly preferential to have this be a string enum since these should all be known types

File: src/models/contracts/objectExplorer/createSessionRequest.ts [hunk 18]
```diff
@@ -0,0 +1,67 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+import {NotificationType, RequestType} from 'vscode-languageclient';
+// tslint:disable-next-line:no-unused-variable
+import { ConnectionDetails } from '../connection';
```
Comment (Charles-Gagnon) [inline src/models/contracts/objectExplorer/createSessionRequest.ts]: Why is this necessary?

File: src/models/contracts/objectExplorer/closeSessionRequest.ts [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+    public errorMessage: string;
+}
+
+
+ // ------------------------------- < Close Session Request > ----------------------------------------------
+
+/**
+ * Establishes an Object Explorer tree session for a specific connection.
+ * This will create a connection to a specific server or database, register
+ * it for use
```
Comment (Charles-Gagnon) [inline src/models/contracts/objectExplorer/closeSessionRequest.ts]: This comment seems wrong

File: src/controllers/untitledSqlDocumentService.ts [hunk 20]
```diff
@@ -16,16 +17,16 @@ export default class UntitledSqlDocumentService {
     /**
      * Creates new untitled document for SQL query and opens in new editor tab
      */
-    public newQuery(): Promise<boolean> {
+    public newQuery(content?: string): Promise<Uri> {
```
Comment (Charles-Gagnon) [inline src/controllers/untitledSqlDocumentService.ts]: >content [](start = 20, length = 7)

Update doc comments

File: src/controllers/mainController.ts [hunk 21]
```diff
@@ -499,11 +540,21 @@ export default class MainController implements vscode.Disposable {
...
+            if (sessionId) {
+                const uri = await this._untitledSqlDocumentService.newQuery(content);
+                // connect to the node if the command came from the context
+                if (!this.connectionManager.isConnected(sessionId)) {
+                    const connectionCreds = this._objectExplorerProvider.getConnectionCredentials(sessionId);
+                    this._statusview.languageFlavorChanged(uri.toString(), Constants.mssqlProviderName);
+                    return this.connectionManager.connect(uri.toString(), connectionCreds);
+                }
+            } else {
+                return this._untitledSqlDocumentService.newQuery().then(uri => {
```
Comment (Charles-Gagnon) [inline src/controllers/mainController.ts]: >then [](start = 67, length = 4)

Just use await here

File: src/controllers/mainController.ts [hunk 22]
```diff
@@ -499,11 +540,21 @@ export default class MainController implements vscode.Disposable {
     /**
      * Opens a new query and creates new connection
      */
-    public onNewQuery(): Promise<boolean> {
+    public async onNewQuery(sessionId?: string, content?: string): Promise<boolean> {
```
Comment (Charles-Gagnon) [inline src/controllers/mainController.ts]: >boolean [](start = 75, length = 7)

Do you want to update this to also return a URI since you did for newQuery()?

File: package.json [hunk 23]
```diff
@@ -154,6 +174,39 @@
...
+          "command": "extension.addObjectExplorer",
+          "title": "%extension.addObjectExplorer%",
+          "group": "navigation"
+        }
+      ],
+      "view/item/context": [
+        {
+          "command": "extension.objectExplorerNewQuery",
+          "when": "viewItem == Server"
+        },
```
Comment (Charles-Gagnon) [inline package.json]: Apparently context keys support OR now [https://github.com/microsoft/vscode/commit/893635042ccd4dbdf919a4493a062330aec2cbc7](https://github.com/microsoft/vscode/commit/893635042ccd4dbdf919a4493a062330aec2cbc7)

File: test/mainController.test.ts (hunk 11 above)
Comment (abist) [reply test/mainController.test.ts]: I'm writing tests now, I haven't completed the PR yet. The next iteration will have them :)

File: src/models/contracts/objectExplorer/refreshSessionRequest.ts (hunk 16 above)
Comment (abist) [reply src/models/contracts/objectExplorer/refreshSessionRequest.ts]: I'm just following what we have in SQL Tools Service, and ADS does the same thing. They both have the same class members so it kind of makes sense.

File: src/models/contracts/objectExplorer/nodeInfo.ts (hunk 17 above)
Comment (abist) [reply src/models/contracts/objectExplorer/nodeInfo.ts]: This comes from the tools service

**Comment by abist** [comment]
@Charles-Gagnon Still not done with the PR. I'll re-request a review once I'm finished with it.

File: src/models/contracts/objectExplorer/createSessionRequest.ts (hunk 18 above)
Comment (abist) [reply src/models/contracts/objectExplorer/createSessionRequest.ts]: Linter otherwise complains with the following error - 
```
ERROR: (no-unused-variable) vscode-mssql/src/models/contracts/objectExplorer/createSessionRequest.ts[8, 1]: All imports on this line are unused.
```

Comment (Charles-Gagnon) [reply src/models/contracts/objectExplorer/createSessionRequest.ts]: Which makes sense since it looks like you aren't actually using ConnectionDetails anywhere. So you should just remove it if that's the case.

Comment (Charles-Gagnon) [reply src/models/contracts/objectExplorer/createSessionRequest.ts]: Oh wait nevermind it's in the generic definition. And actually I recall seeing that issue before...there was a way to fix it though IIRC. Let me look it up - otherwise this is fine

File: src/controllers/mainController.ts (hunk 22 above)
Comment (abist) [reply src/controllers/mainController.ts]: Think it's fine since it doesn't need it

**Comment by abist** [comment]
@Charles-Gagnon merging soon

File: src/controllers/mainController.ts [hunk 24]
```diff
@@ -113,10 +114,52 @@ export default class MainController implements vscode.Disposable {
...
+                    return this._objectExplorerProvider.removeObjectExplorerNode(this._objectExplorerProvider.currentNode);
+                });
+                this.registerCommand(Constants.cmdRefreshObjectExplorerNode);
+                this._event.on(Constants.cmdRefreshObjectExplorerNode, () => {
+                    return this._objectExplorerProvider.refreshNode(this._objectExplorerProvider.currentNode);
+                });
+                this._context.subscriptions.push(vscode.commands.registerCommand(Constants.cmdScriptSelect, async (treeNodeInfo) => {
+                    const objectNames = treeNodeInfo.label.split('.');
+                    const tableName = `[${escapeCharacters(objectNames[0])}].[${escapeCharacters(objectNames[1])}]`;
+                    const databaseName =  treeNodeInfo.parentNode.parentNode.nodeType === Constants.databaseString ?
```
Comment (Charles-Gagnon) [inline src/controllers/mainController.ts]: It might be a good idea to 

1. Check that the current treeNodeInfo.nodeType === Constants.table
2. Instead of assuming that the parent DB is exactly two nodes above, just walk the parents until you find one with nodeType === Constants.database

And then probably display an error if either of those fail (mostly so users are more likely to report the issue and we can investigate)

It's probably not strictly necessary but I really prefer we be safer on things like this since if this did change and start failing for some reason it'd be more annoying to track down without a clear error message.

File: src/controllers/untitledSqlDocumentService.ts [hunk 25]
```diff
@@ -15,19 +16,17 @@ export default class UntitledSqlDocumentService {
...
-                        resolve(true);
-                    });
-                });
+                let doc = await this.vscodeWrapper.openMsSqlTextDocument(content);
+                // Show the new untitled document in the editor's first tab and change the focus to it.
+                await this.vscodeWrapper.showTextDocument(doc, 1, false);
+                resolve(doc.uri);
             } catch (error) {
                 reject(error);
             }
```
Comment (Charles-Gagnon) [inline src/controllers/untitledSqlDocumentService.ts]: With the async you can get rid of the resolve/reject and the whole try/catch

File: package.json (hunk 23 above)
Comment (Charles-Gagnon) [reply package.json]: Not planning on using the new or capabilities? It would be good to try them out to make sure they work. 

---
In reply to: [309266302](https://github.com/microsoft/vscode-mssql/pull/1259#discussion_r309266302) [](ancestors = 309266302)

File: src/constants/constants.ts [hunk 26]
```diff
@@ -55,6 +66,7 @@ export const gettingStartedGuideLink = 'https://aka.ms/mssql-getting-started';
 export const changelogLink = 'https://aka.ms/vscode-mssql-changelog';
 export const integratedAuthHelpLink = 'https://aka.ms/vscode-mssql-integratedauth';
 export const sqlToolsServiceCrashLink = 'https://github.com/Microsoft/vscode-mssql/wiki/SqlToolsService-Known-Issues';
+export const databaseString = 'Database';
```
Comment (Charles-Gagnon) [inline src/constants/constants.ts]: I'd rather see this as a string enum here at least so we know what it's intended use is and makes it easier to discover the nodeTypes that are supported.

**Comment by anthonydresser** [comment]
Late to the party here, but why don't we have icons?

**Comment by abist** [comment]
@anthonydresser Wasn't sure if we wanted to increase the size of the extension just for icons. The extension already has minimal UI, so I was just following that.

**Comment by anthonydresser** [comment]
All the other trees in vscode (at least the onces written by microsoft) have icons, even if its just a subset of icons (maybe server, database, object) its nice from a visual perspective to understand where in the tree you are. @yualan Thoughts?

**Comment by abist** [comment]
@kburtram I added the icons and the MRU functionality. I'll go ahead and merge it tonight if I don't hear about any concerns with the new additions.
